### PR TITLE
MDEV-33971 NAME_CONST in WHERE clause replaced by inner item

### PR DESCRIPTION
**Pull request by DaveGosselin-MariaDB**
Improve performance of queries like
  SELECT * FROM t1 WHERE field = NAME_CONST('a', 4);
by, in this example, replacing the WHERE clause with field = 4 in the case of ref access.

The rewrite is done during fix_fields and we disambiguate this case from other cases of NAME_CONST by inspecting where we are in parsing.  We rely on THD::where to accomplish this.  To improve performance there, we change the type of THD::where to be an enumeration, so we can avoid string comparisons during Item_name_const::fix_fields.  Consequently, this patch also changes all usages of THD::where to conform likewise.

### CODE REVIEW

File: sql/item.cc [hunk 1]
```diff
@@ -2176,6 +2176,11 @@ Item::Type Item_name_const::type() const
 
 bool Item_name_const::fix_fields(THD *thd, Item **ref)
 {
+  if (thd->where == THD_WHERE::WHERE_CLAUSE) {
```
Comment (spetrunia) [inline sql/item.cc:2179]: Coding style: `{` on separate line

File: sql/item.cc [hunk 2]
```diff
@@ -2176,6 +2176,11 @@ Item::Type Item_name_const::type() const
...
+      value_item->fix_fields_if_needed(thd, &value_item);
+      thd->change_item_tree(ref, value_item);
```
Comment (spetrunia) [inline sql/item.cc:2181]: This needs error handling. I think one needs t ocall fix_fields_if_needed for both value_item and name_item and check the result.

I'm wondering why this is needed:

```cpp
 thd->change_item_tree(ref, value_item);
```
did you have anything in mind?

**Review by spetrunia — CHANGES_REQUESTED** [review]
Generally, looks like a step in the right direction. 

I'm wondering if the patch could be smaller if there was a  `const char* thd_where_is_where_cond` and then one could compare `thd->where` strings by comparing pointers... but since we've added an enum, let's not remove it.  It has an advantage of listing the possible values of `thd->where` ...

I think in addition to handling WHERE, we need to handle ON expressions. 
Handling other expressions (like HAVING) looks more risky as they may have references to select list values...

The patch needs test cases... I used this to play:
```

create table t1 (a int, b int);
insert into t1 values (1,1),(2,2);

explain format=json
select * from t1 where a=name_const(1,'varname');

explain format=json
select * from t1 left join t1 as t2 on t1.a=name_const(1,'varname') and t1.b=t2.b;

drop table t1;
```

**Comment by spetrunia** [comment]
One other question:

What if Item_const::fix_fields() always did the following, regardless of its location:

* call `value_item->set_name(...)` in the same way it does for itself in  `Item_const::Item_const`. 
* substitute itself with `value_item`

Will this work and if not why?

**Comment by DaveGosselin-MariaDB** [comment]
> Generally, looks like a step in the right direction.

Thanks!  😄

> I'm wondering if the patch could be smaller if there was a `const char* thd_where_is_where_cond` and then one could compare `thd->where` strings by comparing pointers... but since we've added an enum, let's not remove it. It has an advantage of listing the possible values of `thd->where` ...

I had considered string pointers very briefly, but decided against that approach for a number of reasons:  
 - pointer values are hard to distinguish in logs (should these get logged)
 - enumerations are easier to read and (sometimes) remember
 - future maintainers could look at the string pointer comparisons and say "did he really mean to compare the pointer values and not the strings themselves?" which would lead to endless stumbling. 
 - comparing values across different builds would be impossible

Generally, enumerations are more intentional and less ambiguous.

> I think in addition to handling WHERE, we need to handle ON expressions. Handling other expressions (like HAVING) looks more risky as they may have references to select list values...

Would that be appropriate to do now or as a separate MDEV?

File: sql/item.cc (hunk 2 above)
Comment (DaveGosselin-MariaDB) [reply sql/item.cc:2181]: Without the call to `change_item_tree`, the node for `NAME_CONST` remains in the parse tree.  This call replaces the `NAME_CONST` node with the value it wraps.

**Comment by DaveGosselin-MariaDB** [comment]
> One other question:
> 
> What if Item_const::fix_fields() always did the following, regardless of its location:
> 
> * call `value_item->set_name(...)` in the same way it does for itself in  `Item_const::Item_const`.
> * substitute itself with `value_item`
> 
> Will this work (EDIT: quick attempt: no, rpl.rpl_name_const fails) and if not why?

This is an interesting idea, we should explore this further in another MDEV.

**Review by spetrunia — CHANGES_REQUESTED** [review]
Got a question: 

https://jira.mariadb.org/browse/MDEV-33971?focusedCommentId=285058&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-285058

need to sort this out first.

**Comment by spetrunia** [comment]
I don't understand how the last variant is supposed to work? I see this check

```
if (thd->last_sql_command == SQLCOM_SELECT &&
```
what is it supposed to achieve?

**Comment by spetrunia** [comment]
As far as I understand, the issue raised in 
https://jira.mariadb.org/browse/MDEV-33971?focusedCommentId=285058&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-285058

is still not addressed.
Double-checking this.  I take 
```
commit 05e9b4a6955bcbedc237e46fafdccb2b4419d1a6 (HEAD -> 10.6-mdev-33971, origin/10.6-mdev-33971)
Author: Dave Gosselin <dave.gosselin@mariadb.com>
Date:   Fri Apr 26 12:13:31 2024 -0400

    MDEV-33971 NAME_CONST in WHERE clause replaced by inner item
```
I'm running this testcase:

```
create table t1 (
  a varchar(100) collate utf8_unicode_ci,
  b int
);
insert into t1 values ('foo', 1),('bar', 1);
select * from t1;
 
select * from t1  where a= NAME_CONST('var1',_latin1'foo' COLLATE 'latin1_swedish_ci');

drop table t1;
```

I get:
```
mysqltest: At line 8: query 'select * from t1  where a= NAME_CONST('var1',_latin1'foo' COLLATE 'latin1_swedish_ci')' failed: ER_CANT_AGGREGATE_2COLLATIONS (1267): Illegal mix of collations (utf8mb3_unicode_ci,IMPLICIT) and (latin1_swedish_ci,EXPLICIT) for operation '='
```

Trying the same test on the previous commit:
```
commit 2d5cba22a98e54c0c8975d8d7fe7e04981d73c8a (HEAD)
Author: mariadb-DebarunBanerjee <debarun.banerjee@mariadb.com>
Date:   Thu May 16 13:10:44 2024 +0530
```
I get no error.

File: sql/item.cc [hunk 3]
```diff
@@ -2176,6 +2176,26 @@ Item::Type Item_name_const::type() const
...
+   */
+  if ((thd->where == THD_WHERE::WHERE_CLAUSE ||
+       thd->where == THD_WHERE::ON_CLAUSE) &&
+    !value_item->fix_fields_if_needed(thd, &value_item))
+  {
+    thd->change_item_tree(ref, value_item);
+    if (value_item->collation.derivation != DERIVATION_NUMERIC)
+      value_item->collation.set(value_item->collation.collation,
+                                DERIVATION_IMPLICIT);
+    return FALSE;
```
Comment (spetrunia) [inline sql/item.cc:2195]: The first and obvious question: if we reach this return statement, then
```
    base_flags|= item_base_t::FIXED;
```
is not done for the Item_const object, and item remains unfixed.

File: sql/item.cc [hunk 4]
```diff
@@ -2176,6 +2176,28 @@ Item::Type Item_name_const::type() const
...
+      ... JOIN ... ON foo=NAME_CONST(...)
+    then we have an opportunity to unwrap the NAME_CONST and
+    use the enclosed value directly, replacing NAME_CONST in
+    the parse tree with the value it encloses.
+   */
+  if ((thd->where == THD_WHERE::WHERE_CLAUSE ||
+       thd->where == THD_WHERE::ON_CLAUSE) &&
+      (value_item->type() == FUNC_ITEM ||
+       value_item->type() == CONST_ITEM) &&
+    !value_item->fix_fields_if_needed(thd, &value_item))
```
Comment (spetrunia) [inline sql/item.cc:2191]: We have a call to `value_item->fix_fields_if_needed`  here, and then again we have it below...  I think having all fix_fields calls first and rewrite second would be cleaner...

**Review by spetrunia — CHANGES_REQUESTED** [review]
... Done the above and added a couple of comments.
What do you think of applying this patch: https://gist.github.com/spetrunia/5236479b5c4d5cd286ff00f914c759a9

After that, the patch looks ok to push.

File: sql/item.cc [hunk 5]
```diff
@@ -2176,6 +2176,28 @@ Item::Type Item_name_const::type() const
...
+    If we have either of the following:
+      ... WHERE foo=NAME_CONST(...)
+      ... JOIN ... ON foo=NAME_CONST(...)
+    then we have an opportunity to unwrap the NAME_CONST and
+    use the enclosed value directly, replacing NAME_CONST in
+    the parse tree with the value it encloses.
+   */
+  if ((thd->where == THD_WHERE::WHERE_CLAUSE ||
+       thd->where == THD_WHERE::ON_CLAUSE) &&
+      (value_item->type() == FUNC_ITEM ||
```
[the remark this answers is not in the document]
Comment (DaveGosselin-MariaDB) [reply sql/item.cc:2189]: I don't understand your comment.  This check is necessary to unwrap the NAME_CONST item and replace it in the tree with the item that it wraps (please see the tests).

**Comment by DaveGosselin-MariaDB** [comment]
> ... Done the above and added a couple of comments. What do you think of applying this patch: https://gist.github.com/spetrunia/5236479b5c4d5cd286ff00f914c759a9
> 
> After that, the patch looks ok to push.

OK I will look, thank you.

File: sql/item.cc (hunk 4 above)
Comment (DaveGosselin-MariaDB) [reply sql/item.cc]: Yes I agree, this is nicer, thank you.

**Comment by DaveGosselin-MariaDB** [comment]
> ... Done the above and added a couple of comments. What do you think of applying this patch: https://gist.github.com/spetrunia/5236479b5c4d5cd286ff00f914c759a9
> 
> After that, the patch looks ok to push.

Thank you, this is a nice change and I included it.  Once the tests are done then I will submit, thank you for the OK.

**Pull request merged**